doc/developer/design/20220214_create_cluster.md

```diff
@@ -0,0 +1,191 @@
+an error.
+
+### Index SQL syntax and semantics
+
+The `CREATE INDEX` statement will be extended like so to allow specifying in
+which cluster to create an index:
+
+```
+CREATE INDEX <name> ON <object> (<col>, ...) [IN CLUSTER <cluster>]
+```
+
+If the `IN CLUSTER` clause is not specified explicitly, the index is created in
+the cluster specified by the `cluster` session variable (see below). It is an
+error if that session variable does not name a valid cluster.
+
+Normalizing a `CREATE INDEX` statement will install the explicit `IN CLUSTER`
+clause if it was not explicitly specified.
+
+The syntax of `SHOW INDEXES` will be extended to allow limiting the indexes
+shown to one cluster:
+
+```
+SHOW { INDEX | INDEXES | KEYS } { FROM | IN } <on_name>
+[IN CLUSTER <cluster_name>]
+{ LIKE 'pattern | EXPR <expr> }
+```
+
+The output of `SHOW INDEXES` will be extended with an additional `cluster_name`
+column which displays the name of the cluster in which the index is
+materialized.
+
+### SQL session changes
+
+A new `cluster` session variable will be introduced that specifies the "active"
+cluster for a SQL session. It is analogous to the existing `database` session
+variable.
+
+The default value for the `cluster` session variable will be initially hardcoded
+to `default`. Users can change the value on a per-session basis via the usual
+mechanisms, like the pgwire startup protocol and the `SET` command.
+
+See [future work](#future-work) for details on how we plan to make the default
+configurable in the future.
+
+`SELECT` statements issued in a session will only use the indexes from the
+active cluster.
+
+### System catalog changes
+
+A new `mz_clusters` table with the following structure will describe the
+available clusters in the system:
+
+Field   | Type      | Meaning
+--------|-----------|--------
+id      | `bigint`  | The ID of the cluster.
+name    | `text`    | The name of the cluster.
+virtual | `boolean` | Whether the cluster is virtual.
+size    | `text`    | The size of the cluster.
+
+The existing `mz_indexes` table will be extended with a `cluster_id` field:
+
+Field      | Type      | Meaning
+-----------|-----------|--------
+...        | ...       | ...
+cluster_id | `bigint`  | The ID of the cluster in which the index is materialized.
+
+### Coordinator changes
+
+Internally, each cluster will be identified by an identifier in the range `[1,
+i64::MAX]`. The coordinator will be responsible for allocating identifiers and
+durably recording the mapping from names to identifiers via the following new
+table in the SQLite catalog:
+
+```sql
+CREATE TABLE clusters (
+    id   integer PRIMARY KEY,
+    name text NOT NULL UNIQUE
+);
+```
+
+This table works analogously to the existing `databases` table in the catalog.
+
+The migration that creates the table will additionally install a default row
+of `(1, 'default')`. This creates a cluster named `default` that is present by
```

Comment:
What is the backing of this default cluster? Virtual for anything non-cloud and somehow determined if on cloud?

Reply:
Virtual for everything for now, yeah. A future migration can add the `size` field once we're ready to start implementing that.